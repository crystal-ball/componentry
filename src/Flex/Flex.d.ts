import * as React from 'react'
import { FontProps, SpaceProps } from '../utils/componentry-props'

interface FlexProps {
  /** Sets the `direction` flex property */
  direction?: 'column' | 'column-reverse' | 'row' | 'row-reverse'
  /** Converts component to inline level element */
  inline?: boolean
  /** Sets the `align-items` flex property to control flex items' layout */
  align?: 'start' | 'center' | 'end' | 'baseline' | 'stretch'
  /** Sets the `justify-content` flex property to control flex items' layou */
  justify?: 'start' | 'center' | 'end' | 'around' | 'between' | 'evenly'
  /** Sets the `wrap` flex property to control flex items' wrap behavior */
  wrap?: 'wrap' | 'wrap-reverse' | 'nowrap'
}

/** https://crystal-ball.github.io/componentry/components/buttons */
declare const Flex: React.ComponentType<FlexProps & FontProps & SpaceProps>

export default Flex
