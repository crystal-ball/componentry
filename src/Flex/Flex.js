import { useTheme } from '../Theme/Theme'
import { element } from '../factories/element'

/**
 * [Flex component üìù](https://componentry.design/components/flex)
 */
export function Flex(props) {
  /** @type {{ align: string, direction: string, inline: boolean, justify: string, wrap: string }} */
  const { align, direction, inline, justify, wrap, ...rest } = {
    ...useTheme('Flex'),
    ...props,
  }

  return element({
    componentCx: {
      // ‚ÑπÔ∏è Flex is a helper component so it doesn't have a üÖ≤-flex class
      'd-flex': !inline,
      'd-inline-flex': inline,
      // column, column-reverse, row, row-reverse
      [`flex-${direction}`]: direction,
      // wrap, wrap-reverse, nowrap
      [`flex-${wrap}`]: wrap,
      // start, end, center, baseline, stretch
      [`align-items-${align}`]: align,
      // start, end, center, around, between, evenly
      [`justify-content-${justify}`]: justify,
    },
    ...rest,
  })
}
Flex.displayName = 'Flex'
