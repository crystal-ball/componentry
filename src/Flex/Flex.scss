/* ~componentry/src/Flex/Flex.scss */

/// Componentry Flex styles that don't map to a Tailwind utility style are
/// included as a component style to ensure they're available for users that
/// switch to Tailwind for their utility styles

@each $breakpoint in map-keys(theme-get('breakpoints')) {
  @include media-breakpoint-up($breakpoint) {
    $infix: breakpoint-infix($breakpoint, theme-get('breakpoints'));

    // Provide "flex-column" and "flex-column-reverse" classes to map to Flex's
    // direction prop values.
    // (Tailwind provides "flex-col" and "flex-col-reverse", though short these
    // don't map to the CSS styles and Componentry prioritizes consistent
    // intuition of values over brevity)

    .#{$infix}flex-column {
      flex-direction: column !important;
    }

    .#{$infix}flex-column-reverse {
      flex-direction: column-reverse !important;
    }
  }
}
