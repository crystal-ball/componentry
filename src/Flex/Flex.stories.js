/* global module */

import React from 'react'
import { storiesOf } from '@storybook/react'
import { boolean, select } from '@storybook/addon-knobs'

import Flex from './Flex'

const directions = {
  column: 'column',
  'column-reverse': 'column-reverse',
  row: 'row',
  'row-reverse': 'row-reverse',
  None: null,
}

const wrap = {
  wrap: 'wrap',
  'wrap-reverse': 'wrap-reverse',
  nowrap: 'nowrap',
  None: null,
}

const align = {
  start: 'start',
  end: 'end',
  center: 'center',
  baseline: 'baseline',
  stretch: 'stretch',
  None: null,
}

const justify = {
  start: 'start',
  end: 'end',
  center: 'center',
  around: 'around',
  between: 'between',
  evenly: 'evenly',
  None: null,
}

storiesOf('Flex', module).add('<Flex />', () => (
  <div className='w-100 min-100vh' style={{ height: 1 }}>
    <Flex
      className='w-100 h-100'
      inline={boolean('Inline', false)}
      direction={select('Direction', directions, null)}
      wrap={select('Wrap', wrap, null)}
      align={select('Align', align, null)}
      justify={select('Justify', justify, null)}
    >
      {[1, 2, 3].map(key => (
        <div className='border' key={key}>
          Flex content #{key}
        </div>
      ))}
    </Flex>
  </div>
))
