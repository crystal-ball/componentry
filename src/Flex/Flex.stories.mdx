import { boolean, select } from '@storybook/addon-knobs'

import Flex from './Flex'

<Meta title='Components|Flex' component={Flex} />

<DocsTitle title="Flex" experimental/>

## Block containers

Flex components create block level flex containers:

<Flex>
  {[1, 2, 3].map(key => (
    <div className='border' key={key}>
      Flex content #{key}
    </div>
  ))}
</Flex>

## Inline containers

Create inline level flex containers using the `inline` prop:

<Flex inline>
  {[1, 2].map(key => (
    <div className='border' key={key}>
      Flex content #{key}
    </div>
  ))}
</Flex>
<Flex inline>
  {[1, 2].map(key => (
    <div className='border' key={key}>
      Flex content #{key}
    </div>
  ))}
</Flex>

## Flex direction

Override the default direction of `row` with the `direction` prop:

<Flex direction='column'>
  {[1, 2, 3].map(key => (
    <div className='border' key={key}>
      Flex content #{key}
    </div>
  ))}
</Flex>

## Justify and align

Set the flex items justification and alignment with the `justify` and `align` props:

<Flex justify="center" align="center">
  {[1, 2, 3].map(key => (
    <div className='border' key={key}>
      Flex content #{key}
    </div>
  ))}
</Flex>

## Flex wrap

Override the flex wrap behavior with the `wrap` props:

<Flex wrap='no-wrap'>
  {[1, 2, 3].map(key => (
    <div className='border' key={key}>
      Flex content #{key}
    </div>
  ))}
</Flex>
