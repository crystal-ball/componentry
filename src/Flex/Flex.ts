import React from 'react'
import { useTheme } from '../Theme/Theme'
import { ComponentBaseProps } from '../utils/types'
import { element } from '../utils/element-creator'

interface FlexProps extends ComponentBaseProps<'div'> {
  /** Sets an `align-items` flex style */
  align?: 'start' | 'center' | 'end' | 'stretch' | 'baseline'
  /** sets a `flex-direction` flex style */
  direction?: 'row' | 'column' | 'row-reverse' | 'column-reverse'
  /** Switches between display between an inline and block element */
  inline?: boolean
  /** Sets a `justify-content` flex style */
  justify?: 'start' | 'center' | 'end' | 'between' | 'around' | 'evenly'
  /** Sets a `flex-wrap` flex style */
  wrap?: 'wrap' | 'nowrap' | 'wrap-reverse'
}

/**
 * [Flex component üìù](https://componentry.design/components/flex)
 */
export const Flex: React.FC<FlexProps> = (props) => {
  const { align, direction, inline, justify, wrap, ...rest } = {
    ...useTheme<FlexProps>('Flex'),
    ...props,
  }

  return element('Flex', {
    componentCx: {
      'd-flex': !inline,
      'd-inline-flex': inline,
      [`flex-${direction}`]: direction,
      [`flex-${wrap}`]: wrap,
      [`align-items-${align}`]: align,
      [`justify-content-${justify}`]: justify,
    },
    ...rest,
  })
}
Flex.displayName = 'Flex'
