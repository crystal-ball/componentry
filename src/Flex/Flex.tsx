import React from 'react'
import { useTheme } from '../Theme/Theme'
import { element } from '../factories/element'
import { BaseProps } from '../utils/base-types'

interface FlexProps
  extends BaseProps,
    Omit<React.ComponentPropsWithoutRef<'div'>, 'className'> {
  /** Sets an `align-items` flex style */
  align?: 'start' | 'center' | 'end' | 'stretch' | 'baseline'
  /** sets a `flex-direction` flex style */
  direction?: 'row' | 'column' | 'row-reverse' | 'column-reverse'
  /** Switches between display between an inline and block element */
  inline?: boolean
  /** Sets a `justify-content` flex style */
  justify?: 'start' | 'center' | 'end' | 'between' | 'around' | 'evenly'
  /** Sets a `flex-wrap` flex style */
  wrap?: 'wrap' | 'nowrap' | 'wrap-reverse'
}

/**
 * [Flex component üìù](https://componentry.design/components/flex)
 */
export const Flex: React.FC<FlexProps> = (props) => {
  const { align, direction, inline = false, justify, wrap, ...rest } = {
    ...useTheme<FlexProps>('Flex'),
    ...props,
  }

  return element({
    componentCx: {
      // ‚ÑπÔ∏è Flex is a helper component so it doesn't have a üÖ≤-flex class
      'd-flex': !inline,
      'd-inline-flex': inline,
      // column, column-reverse, row, row-reverse
      [`flex-${direction}`]: direction,
      // wrap, wrap-reverse, nowrap
      [`flex-${wrap}`]: wrap,
      // start, end, center, baseline, stretch
      [`align-items-${align}`]: align,
      // start, end, center, around, between, evenly
      [`justify-content-${justify}`]: justify,
    },
    ...rest,
  })
}
Flex.displayName = 'Flex'
