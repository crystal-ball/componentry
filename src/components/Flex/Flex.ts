import React from 'react'
import { useTheme } from '../Theme/Theme'
import { ComponentBaseProps } from '../../utils/types'
import { element } from '../../utils/element-creator'

export interface FlexProps extends ComponentBaseProps<'div'> {
  /** Sets an `align-items` style */
  align?: 'start' | 'end' | 'center' | 'baseline' | 'stretch'
  /** Sets a `flex-direction` flex style */
  direction?: 'column' | 'column-reverse' | 'row-reverse' | 'row'
  /** Switches between display between an inline and block element */
  inline?: boolean
  /** Sets a `justify-content` style */
  justify?: 'start' | 'end' | 'center' | 'between' | 'around' | 'evenly'
  /** Sets a `flex-wrap` flex style */
  wrap?: 'wrap' | 'nowrap' | 'wrap-reverse'
}

/**
 * [Flex component üìù](https://componentry.design/components/flex)
 */
export const Flex: React.FC<FlexProps> = (props) => {
  const { align, direction, inline, justify, wrap, ...rest } = {
    ...useTheme<FlexProps>('Flex'),
    ...props,
  }
  // Tailwind uses a flex-col class but direction="col" is super wonky
  // => so props use "column" and we replace with "col"
  const computedDirection = direction?.replace('column', 'col')

  return element({
    componentCx: {
      'flex': !inline,
      'inline-flex': inline,
      [`flex-${computedDirection}`]: computedDirection,
      [`flex-${wrap}`]: wrap,
    },
    alignItems: align,
    justifyContent: justify,
    ...rest,
  })
}
Flex.displayName = 'Flex'
