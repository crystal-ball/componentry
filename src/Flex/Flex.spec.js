import React from 'react'
import { render, screen } from '@testing-library/react'

import Flex from './Flex'

import elementTests from '../../test/element-tests'

describe('<Flex />', () => {
  elementTests(Flex)

  test('When inline is not passed, then d-flex className is rendered', () => {
    render(<Flex>Content</Flex>)
    expect(screen.getByText('Content')).toHaveClass('d-flex')
  })

  test('When inline is passed, then d-inline-flex className is rendered', () => {
    render(<Flex inline>Content</Flex>)
    expect(screen.getByText('Content')).toHaveClass('d-inline-flex')
  })

  test('When modifier props are passed, then the expanded classNames are rendered', () => {
    render(
      <Flex align='start' direction='column' justify='start' wrap='wrap'>
        Content
      </Flex>,
    )
    expect(screen.getByText('Content')).toHaveClass(
      'd-flex flex-column align-items-start flex-wrap justify-content-start',
    )
  })
})

// Snapshots
// ---------------------------------------------------------------------------
describe('<Flex /> snapshots', () => {
  it('renders correctly', () => {
    const { container } = render(
      <div>
        <Flex>Flex content</Flex>
        <Flex inline>Inline flex content</Flex>
      </div>,
    )

    expect(container.firstChild).toMatchSnapshot()
  })
})
