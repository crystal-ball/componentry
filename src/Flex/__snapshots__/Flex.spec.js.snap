// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Flex /> snapshots renders correctly 1`] = `
<div
  class="flex"
  data-testid="flex"
>
  Flex content
</div>
`;
